### ISSUE TITLE
Updating SublimeLinter should just work, no?

### ISSUE THREAD

**Issue by kaste** [SublimeLinter/SublimeLinter#991]
It seems that every (silent) update breaks SublimeLinter and requires a restart. I guess you, @FichteFoll , should know everything about this. 😄 I know nothing about this and am a bit disappointed that PackageControl can't just do it. What to do?

- Toast the user after every update, that a restart is required
- Fix the underlying issue

**Comment by braver** [comment]
There is an event we can hook into. The problem is successfully reloading all our modules, and after that probably also reloading all the plugins. Reloading neatly after an upgrade with a plugin like ours, in an environment like ST, is probably impossible to solve efficiently and reliably.

We can just hook into the upgrade event and show our message panel?

**Comment by FichteFoll** [comment]
Related issue in PC: https://github.com/wbond/package_control/issues/1325

As I mentioned in some other issue, for PackageDev I added a manual cleanup of `sys.modules` to clear up any lingering references and importing them causes a reload of the modules. Currently, there is still an issue with `_logging.py` being removed as well, but that shouldn't matter for ST since nothing imports that module and ST handles its loaded plugins manually anyway. I don't think I ever tested this, but in theory it "might just work". For your package, you just need to ensure that all your state in submodules (e.g. the linter daemon or ThreadExecutors) are cleaned up when `plugin_unloaded` is called, which happens when PC initiates the update (or when the package is uninstalled/disabled).

https://github.com/SublimeText/PackageDev/blob/dbac9c8334ba02706c39151d23ca971c5602126e/main.py

**Comment by FichteFoll** [comment]
The most problematic thing for SL however is that other external packages, i.e. linter plugins, may have already imported the base linter class and subclassed it and by reloading the SL main modules, you lose those subclasses registered with the metaclass. There are ways to persist the classes that have been registered *but* they will still have subclassed the *old* Linter classes and all their super() method calls will be resolved to the old Linter classes instead of the new reloaded ones.

Basically, having to restart ST seems like the only option here.

**Comment by braver** [comment]
Ok, so two options:

1. Lots of effort to get SL to reload properly, plus try to reload the plugins, and it "might work", ending up telling the user to restart anyway to be on the safe side
2. Spend less time than typing this thread and just tell the user to restart

I have a preference I think ;)

**Comment by FichteFoll** [comment]
>How do I disable/enable forcefully other plugins,

Load `Preferences.sublime-settings` and modify the `ignored_packages` list.

**Comment by braver** [comment]
Oi, that's super messy move.

**Comment by kaste** [comment]
That's a real hack. I'll skip that one.

**Comment by FichteFoll** [comment]
Yeah, just telling users to restart is the easiest approach.

I guess you could just refuse to expose any plugin functionality or something after an update so that users don't run into undefined behavior through partial upgrades in their environment, in addition to a message of course.
